### PR TITLE
prometheus: Added package prometheus-slurm-exporter

### PR DESCRIPTION
**Pull request by Rovanion**
###### Motivation for this change
Added the package prometheus-slurm-exporter. A program for exposing statistics about the Slurm cluster scheduler to the Prometheus database.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -7891,6 +7891,12 @@
     githubId = 3621083;
     name = "Roosembert (Roosemberth) Palacios";
   };
+  rovanion = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:7894]: Please put this into a separate commit with the message ``maintainers: add rovanion``.

File: pkgs/servers/monitoring/prometheus/slurm-exporter.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{ lib, buildGoPackage, fetchFromGitHub, nixosTests }:
+
+buildGoPackage rec {
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/slurm-exporter.nix:3]: Please use buildGoModule.

File: pkgs/servers/monitoring/prometheus/slurm-exporter.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  pname = "slurm-exporter";
+  version = "0.14";
+
+  goPackagePath = "github.com/vpenso/prometheus-slurm-exporter";
+
+  src = fetchFromGitHub {
+    rev = version;
+    owner = "vpenso";
+    repo = "prometheus-slurm-exporter";
+    sha256 = "02d71q0pzgidn63x6hqz48kih3qw3dqlmwb09wgrcahdvjhrviqq";
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/slurm-exporter.nix:13]:
```suggestion
    owner = "vpenso";
    repo = "prometheus-slurm-exporter";
    rev = version;
    sha256 = "02d71q0pzgidn63x6hqz48kih3qw3dqlmwb09wgrcahdvjhrviqq";
```

File: pkgs/servers/monitoring/prometheus/slurm-exporter.nix [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  };
+
+  goDeps = ./slurm-exporter-deps.nix;
+
+  meta = with lib; {
+    description = "A Prometheus stats exporter for the Slurm cluster manager. ";
+    homepage = "https://github.com/vpenso/prometheus-slurm-exporter";
+    license = licenses.gpl3;
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/slurm-exporter.nix:21]: gpl3Plus or gpl3Only?

File: pkgs/servers/monitoring/prometheus/slurm-exporter.nix [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    maintainers = with maintainers; [ rovanion ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/slurm-exporter.nix:23]:
```suggestion
```

Set by buildGoModule.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by Rovanion** [comment]
I think that I have fixed all your comments @SuperSandro2000, but now the package fails in the test phase due to a binary `squeue` not being present when the tests are run. This binary is part of Slurm, is there a way I can make it available during the tests or should they be disabled?

File: pkgs/servers/monitoring/prometheus/slurm-exporter.nix [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  src = fetchFromGitHub {
+    owner = "vpenso";
+    repo = "prometheus-slurm-exporter";
+    rev = version;
+    sha256 = "02d71q0pzgidn63x6hqz48kih3qw3dqlmwb09wgrcahdvjhrviqq";
+  };
+
+  goDeps = ./slurm-exporter-deps.nix;
+
```
Comment (SuperSandro2000) [inline pkgs/servers/monitoring/prometheus/slurm-exporter.nix:15]:
```suggestion

```
And the file itself can be deleted.

**Review by SuperSandro2000 — COMMENTED** [review]
> This binary is part of Slurm, is there a way I can make it available during the tests or should they be disabled?

Try checkInputs. It puts binaries on the PATH during testing.

**Comment by Rovanion** [comment]
I tried adding:

```
checkInputs = [ slurm-full ];
```

or variations thereof in the package definition but to no avail. The thing I've been fighting with today is getting the test phase to output anything more informative than that it failed with an exit code 1. I'll return tomorrow and explore more.

**Comment by Rovanion** [comment]
Figured out why the tests were failing. Turns out that they need a working Slurm controller configured to run. So I disabled the tests. Also removed the requested line and file about transitive dependencies.

If there is a way to make a service with a specific config available during the test phase I could try that.

**Comment by bjornfor** [comment]
@Rovanion Take a look at the nixos test infrastructure.

**Comment by Rovanion** [comment]
I have tried figuring out what the nixos test infrastructure is today and have come to the conclusion that it is a set of system integration tests run after the package builds are complete, rather than something which augments the checkPhase. Am I correct in this understanding?

**Comment by bjornfor** [comment]
> I have tried figuring out what the nixos test infrastructure is today and have come to the conclusion that it is a set of system integration tests run after the package builds are complete, rather than something which augments the checkPhase. Am I correct in this understanding?
> 
> Put in other words: It will not make a Slurm cluster be available during the checkPhase. Rather, its something that can be added instead of the checkPhase.

Correct.

**Comment by Rovanion** [comment]
Hi again all!
I have tried my best to set up both unit tests and system tests for prometheus-slurm-exporter. In doing the former I discovered that prometheus-slurm-exporter had no actual unit tests, they only ever printed the output of a parser function without checking that the output was what was expected. So after splitting up the unit tests and system tests of prometheus-slurm-exporter I wrote unit tests for all the parser functions interfacing with the Slurm commands. I also wrote system tests which invokes the actual system commands. These are now available in `make unittest` and `make systemtest` of the project.

Being done with that I've now tried my hands at setting up NixOS system tests again. But I'm now stuck where nix-specific knowledge is needed, and I would like to ask for help: The problem is that the Slurm commands, like `sinfo` and `sacct`, need `/etc/slurm.conf` for them to work. But this config file is not present after the Slurm service has been installed on the machine.

Looking at the source code that defines the Slurm service `nixos/modules/services/computing/slurm/slurm.nix` it seems like the Slurm commands are put in a wrapper that prepends `SLURM_CONF="${etcSlurm}/slurm.conf"` to each of the command invocations. Now this obviously works for when `sinfo` is run from a shell that has the `PATH` set up like Nix wants it to, but the Go service prometheus-slurm-exporter seems to be missing this.

So does anyone know how to set up the `PATH` for Go processes, or how to get the `${etcSlurm}` path so that I can patch the prometheus-slurm-exporter source to prepend it to all subprocess invocations, or some other way to address this issue? Because right now the `prometheus-slurm-exporter` fails to gather data because the subprocesses fail to find the Slurm configuration.

And another question: Is it possible to find the source code of the `prometheus-slurm-exporter` during the NixOS testing infrastructure run and run `make systemtest`?

**Comment by Rovanion** [comment]
@SuperSandro2000 Is there anything in your review that I have not fixed so far?

**Review by markuskowa — APPROVED** [review]
@SuperSandro2000 this looks good to me now. Is there anything else or should we just merge it.

**Comment by markuskowa** [comment]
@Rovanion would you mind rebasing the PR against master? I would then just go ahead and merge it to have it in NixOS 22.05

**Comment by Rovanion** [comment]
I'm afraid I have moved on and currently have no infrastructure running Nix to test on.
